pkg/provider/kubernetes/crd/kubernetes.go

```diff
@@ -507,6 +508,52 @@ func buildTLSOptions(ctx context.Context, client Client) map[string]tls.Options
 	return tlsOptions
 }
 
+func buildTLSStores(ctx context.Context, client Client) map[string]tls.Store {
+	tlsStoreCRD := client.GetTLSStores()
+	var tlsStores map[string]tls.Store
+
+	if len(tlsStoreCRD) == 0 {
+		return tlsStores
+	}
+	tlsStores = make(map[string]tls.Store)
+
+	for _, tlsStore := range tlsStoreCRD {
+		namespace := tlsStore.Namespace
+		logger := log.FromContext(log.With(ctx, log.Str("tlsStore", tlsStore.Name), log.Str("namespace", namespace)))
+		secretName := tlsStore.Spec.DefaultCertificate.SecretName
+
+		secret, exists, err := client.GetSecret(namespace, secretName)
+		if err != nil {
+			logger.Errorf("Failed to fetch secret %s/%s: %v", namespace, secretName, err)
+			continue
+		}
+		if !exists {
+			logger.Errorf("Secret %s/%s does not exist", namespace, secretName)
+			continue
+		}
+
+		cert, key, err := getCertificateBlocks(secret, namespace, secretName)
+		if err != nil {
+			logger.Errorf("Could not get certificate blocks: %v", err)
+			continue
+		}
+
+		if _, ok := tlsStores["default"]; ok {
+			break
+		}
+
+		// Since only the default TLS Store is used, we just need to set it if it doesn't exist.
+		tlsStores["default"] = tls.Store{
+			DefaultCertificate: &tls.Certificate{
+				CertFile: tls.FileOrContent(cert),
+				KeyFile:  tls.FileOrContent(key),
+			},
+		}
```

Comment:
So, my last remark was under the current algorithm assumption, i.e. that we only wanted to keep whatever first store we find and call it the "default" one. But after chatting with the team, I don't think that's what we want to do.

We need to make sure that the day we start supporting stores other than the default one, we do not break the pre-existing behaviour for a given config/user-provided CRD. I think that means it's not the role of the config builder here to make the call about whether or not a TLS store is a useful one (aka the default one). In other words, each of them should be kept, with:

    tlsStores[tlsStore.Name] = blabla

and then it's up to the route handlers to pick up the tlsStore they need/want in that map later on. even though we know that for now they'll only ever pick the tlsStores["default"] on).

If in doubt, I think you can look at how things behave in pkg/provider/file/file.go on that matter.

Reply:
The issue here is: What do we take the store name to be?

Do we take it to be the name of the kubernetes CRD object? (note that it will more than likely not be "default", or there could be a "default" in every namespace).

Reply:
well yeah, I don't see what other choice we have than the object name. Is there any other possibility/suggestion?

Reply:
Not really without making any changes.

Ideally we use the same `makeID` as the TLSOptions, but as soon as you add the namespace into the mix, it will never make a "default" key.

I think this is the best option for now.